src/ape/managers/chain.py

```diff
@@ -975,71 +980,100 @@ def get(
-            if default and default != contract_type:
-                # Replacing contract type
-                self._local_contract_types[address_key] = default
-                return default
-
-            return contract_type
 
+        # Handle the local network case first
         if self._network.name == LOCAL_NETWORK_NAME:
             # Don't check disk-cache or explorer when using local
             if default:
                 self._local_contract_types[address_key] = default
 
             return default
 
-        contract_type = self._get_contract_type_from_disk(address_key)
-        if not contract_type:
+        # Attempt to get the contract type from memory
+        contract_type_from_memory = self._local_contract_types.get(address_key)
+
+        # Attempt to get the contract type from disk
+        contract_type_from_disk = self._get_contract_type_from_disk(address_key)
+
+        # Handle the case where no contract is cached on disk
+        if contract_type_from_disk is None:
             # Contract could be a minimal proxy
             proxy_info = self._local_proxies.get(address_key) or self._get_proxy_info_from_disk(
                 address_key
             )
 
+            # Attempt to retrieve missing proxy info
             if not proxy_info:
                 proxy_info = self.provider.network.ecosystem.get_proxy_info(address_key)
-                if proxy_info and self._is_live_network:
+                if proxy_info and self._is_live_network and cache_to_disk:
+                    # If found, cache to disk
                     self._cache_proxy_info_to_disk(address_key, proxy_info)
 
+            # If proxy info is already known, save the proxy info to memory and
+            # call this function again at the implementation contract address
             if proxy_info:
                 self._local_proxies[address_key] = proxy_info
-                return self.get(proxy_info.target, default=default)
+                return self.get(proxy_info.target, default=default, cache_to_disk=cache_to_disk)
 
             if not self.provider.get_code(address_key):
+                # If no code is stored at this address, store the provided default contract
+                # type in memory. This block is skipped if a contract type was not provided
+                # (`default==None`)
                 if default:
                     self._local_contract_types[address_key] = default
-                    self._cache_contract_to_disk(address_key, default)
+                    if cache_to_disk:
+                        self._cache_contract_to_disk(address_key, default)
 
                 return default
 
-            # Also gets cached to disk for faster lookup next time.
-            contract_type = self._get_contract_type_from_explorer(address_key)
+        # Attempt to get the contract type from block explorer
+        # (unless overridden by `fetch_from_explorer=False`)
+        contract_type_from_explorer = (
```

Comment:
this defeats the purpose of caching, doesn't it?
the explorer API method is hit every time with these changes.

should `fetch_from_explorer` default to `False` if the contract type was not found on disk and `True` otherwise? That would safe an API call in most cases I believe.

Reply:
Wonder if there's some way to validate that an ABI is "fully specified", which would bypass this check because we know we can't expand it anymore to be "more" correct

Would also have to account for  proxymutability in this scenario too

Reply:
Gives me ideas for refactoring to the `QueryAPI` system. Basically just fetch the contract type from any available source (cache, etherscan, sourcity, etc.) and merges the results into the best possible ABI

If etherscan is rate limiting, just have it bypass. If the cache was set to a lesser ABI, extend it with info from external sources, etc.

Reply:
if etherscan had a webhook for contract source updates, we could theoretically automatically update caches in the background.

Reply:
> this defeats the purpose of caching, doesn't it? the explorer API method is hit every time with these changes.
The case I'm thinking of is memory or disk has cached some minimal ABI, but you want to update it from the block explorer. `get` is called through `instance_at`, where I've added a ternary statement that disables the block explorer fetch if a contract type was passed via the `default` argument. 
 
> should `fetch_from_explorer` default to `False` if the contract type was not found on disk and `True` otherwise? That would safe an API call in most cases I believe.
If `cache_to_disk` was set to `True` the first time a contract was loaded, the disk cache will be empty on subsequent runs. If we wanted to update the disk cache later, I think we should discount the state of whatever is in the disk cache (missing or outdated).